crates/bevy_tasks/src/task_pool.rs

```diff
@@ -266,41 +267,76 @@ impl TaskPool {
     ///         });
     ///     });
     /// }
-    ///
     pub fn scope<'env, F, T>(&self, f: F) -> Vec<T>
     where
         F: for<'scope> FnOnce(&'scope Scope<'scope, 'env, T>),
         T: Send + 'static,
     {
-        Self::THREAD_EXECUTOR.with(|thread_executor| {
-            // SAFETY: This safety comment applies to all references transmuted to 'env.
-            // Any futures spawned with these references need to return before this function completes.
-            // This is guaranteed because we drive all the futures spawned onto the Scope
-            // to completion in this function. However, rust has no way of knowing this so we
-            // transmute the lifetimes to 'env here to appease the compiler as it is unable to validate safety.
-            let executor: &async_executor::Executor = &self.executor;
-            let executor: &'env async_executor::Executor = unsafe { mem::transmute(executor) };
-            let thread_executor: &'env ThreadExecutor<'env> =
-                unsafe { mem::transmute(thread_executor) };
-            let spawned: ConcurrentQueue<FallibleTask<T>> = ConcurrentQueue::unbounded();
-            let spawned_ref: &'env ConcurrentQueue<FallibleTask<T>> =
-                unsafe { mem::transmute(&spawned) };
-
-            let scope = Scope {
-                executor,
-                thread_executor,
-                spawned: spawned_ref,
-                scope: PhantomData,
-                env: PhantomData,
-            };
-
-            let scope_ref: &'env Scope<'_, 'env, T> = unsafe { mem::transmute(&scope) };
-
-            f(scope_ref);
-
-            if spawned.is_empty() {
-                Vec::new()
-            } else {
+        Self::THREAD_EXECUTOR
+            .with(|thread_executor| self.scope_with_executor_inner(true, thread_executor, f))
+    }
+
+    /// This allows passing an external executor to spawn tasks on. When you pass an external executor
+    /// [`Scope::spawn_on_scope`] spawns is then run on the thread that [`ThreadExecutor`] is being ticked on.
+    /// See [`Self::scope`] for more details in general about how scopes work.
+    pub fn scope_with_executor<'env, F, T>(
+        &self,
+        tick_task_pool_executor: bool,
+        thread_executor: Option<&ThreadExecutor>,
+        f: F,
+    ) -> Vec<T>
+    where
+        F: for<'scope> FnOnce(&'scope Scope<'scope, 'env, T>),
+        T: Send + 'static,
+    {
+        if let Some(thread_executor) = thread_executor {
```

Comment:
This logic is hard to follow at first pass; I'd like a comment on this if/else block.